python/plugins/processing/algs/qgis/Datasources2Vrt.py

```diff
@@ -125,11 +127,17 @@ def mergeDataSources2Vrt(self, dataSources, outFile, union=False, relative=False
 
             feedback.setProgress(int(current * total))
 
-            inFile = layer.source()
+            inFile = layer.source()            
             srcDS = ogr.Open(inFile, 0)
             if srcDS is None:
-                raise QgsProcessingException(
-                    self.tr('Invalid datasource: {}'.format(inFile)))
+                #might be a shapefile loaded as a directory of shapefiles. We try to resolve the uri to a filepath
+                sourceParts = QgsProviderRegistry.instance().decodeUri('ogr', inFile)
+                if sourceParts.get("layerName"):
+                    inFile = os.path.join(sourceParts.get("path"), sourceParts.get("layerName") + ".shp")
```

Comment:
Won't this break if a geopackage layer (or similar) was loaded? In that case the layerName parameter will be present, but the file name generated here will be broken.

This is really quite tricky to handle -- I think we'd need a check in here to first see if sourceParts['path'] is actually a directory (and if not, there's nothing required). Then, if it IS a directory, we'd need to check inside that path for files which start with the layername and have known vector data format extensions (because the same situation may arise for say KML files), and use that to determine the actual name of the file. Or alternatively, maybe we could use OGR api to try to load the file using just the path and layername, and let it run whatever internal logic it has to resolve the layername to a file, and then retrieve this actual file name from the OGR layer...

Reply:
It wont actually break anything, since gpkgs dont work now. Any layer with a source that is not an actual filepath cannot be used with that algorithm at the moment. However, I agree that the trick i added is not very general and only fixes the case of shapefiles.

I improved it a bit and put it in a separate function in the processing.tools.vector module. From there, it can be moved to C++ if someone takes care of that.

Still, some cases are not handled. It works now for gpkgs, but as long as the layer is the only one in the gpkg file. 

In short, it's an improvement over the previous status of the algorithm, but it wont work with all input layers, like other Processing algorithms do. Other algorithms might be affected of this issue, so it's worth checking and maybe adding a similar mechanism.